mautic/mautic
# PHPSTAN analysis on CampaignBundle

[//]: # ( Please answer the following questions: )

| Q  | A
| --- | ---
| Bug fix? | N
| New feature? | N
| Related user documentation PR URL | /
| Related developer documentation PR URL | In the [README.md](https://github.com/mautic/mautic/pull/3952/files#diff-04c6e90faac2675aa89e2176d2eec7d8)
| Issues addressed (#s or URLs) | /
| BC breaks? | N
| Deprecations? | N

[//]: # ( Note that all new features should have a related user and/or developer documentation PR in their respective repositories. )

[//]: # ( Required: )
#### Description:

[PHPSTAN](https://github.com/phpstan/phpstan) is a library which creates static analysis of PHP code. It can help us to find issues before our users do. 

We cannot analyse the whole Mautic codebase right now since there are too many issues which PHPSTAN uncovers. The intention behind this PR is to start small - fix only CampaignBundle and add it to automatic tests to ensure the new code won't introduce new issues. Then add another bundle and so on.

One downside of PHPSTAN is that it supports only PHP 7+, so `composer require  phpstan/phpstan` will fail on PHP 5.6, so we have to install this package globally and use it only on PHP7+. See the [README.md](https://github.com/mautic/mautic/pull/3952/files#diff-04c6e90faac2675aa89e2176d2eec7d8) changes for more details.

#### Steps to test this PR:
1. Apply this PR
2. Run `composer test` to make sure all passes. Also 2 more unit tests were added.
3. Click through the UI. There were quite a few changes needed to make the analysis happy.
4. Run phpunit on PHP API Library to make sure it didn't break any functionality there. For example the stats endpoint has been modified a bit.
